NixOS module support: Separate NUR nixpkgs from repos nixpkgs, avoid callPackages

Infinisil opened the pull request:
This is essential to get modules in NUR to work. By taking a separate
argument for NUR's nixpkgs (for fetchgit, fetchzip and lib), we don't
need to evaluate the nixpkgs used for repos.

This also implies that you won't be able to `callPackage` NUR anymore,
and instead you'll have to use `import (builtins.fetchGit ".../NUR") {
inherit pkgs; }` instead. Doing this also prevents the evaluation of
pkgs. In case of NixOS, this pkgs depends on your whole config, which is
the source of the recursion. Evaluating this at the last possible moment
is key.

This also means that you won't be able to take package arguments in a
repo definition, you instead get just `pkgs`, also to avoid evaluation
of it.

New install instructions would be
```nix
{
  packageOverrides = pkgs: {
    nur = import (builtins.fetchGit "https://github.com/nix-community/NUR") {
      inherit pkgs;
    };
  };
}
```

To get NixOS modules to work you'll have to use something like
```nix
{ pkgs, ... }:
let
  nur = import (builtins.fetchGit "https://github.com/nix-community/NUR") {
    inherit pkgs;
  };
in {
  imports = [
    nur.repos.infinisil.module
  ];
}
```

Preferably one would set `nur=https://github.com/nix-community/NUR` in their NIX_PATH, which allows one to do
```nix
{ pkgs, ... }: {
  imports = [
    (import <nur> { inherit pkgs; }).repos.infinisil.module
  ];
}
```

Also the pkgs argument should be made optional, so you don't even need to pass it when importing a module.

Docs and stuff would need to be updated, I might do this later.

Mic92 commented:
I like the idea of having a separate expression source for fetchgit and co. I try to think about the most straightforward syntax the next days or so. Documentation should be definitely added before merging any changes.

Mic92 commented:
One thing that sucks about `fetchgit` or `fetchTarball` is that it will not fallback to a cached version when the ttl expires. I could not change my timezone without internet that way with nur modules in my list. Maybe a channel like behavior is more appropriate.

dtzWill commented:
:+1: to channel-like behavior, and if not anything that can fallback when remote resource is unavailable (server down, you're on a plane, etc.) would be great. Kinda lame to have eval fail in those cases.  I filed an issue on Nix about this and one solution suggested was to basically (temporarily) bump ttl to infinity (or some high value).  Hope this helps and regardless glad your considering this with NUR!

Infinisil commented:
In the latest commit I added support for still being able to use the deprecated callPackage syntax, but it issuing a warning. Now when you try to use such a repository for a NixOS module you'll get this:
```
trace: WARNING: NUR repository infinisil is using the deprecated callPackage syntax which
will result in infinite recursion when used with NixOS modules
error: infinite recursion encountered, at /home/infinisil/src/nixpkgs/lib/modules.nix:163:28
(use '--show-trace' to show detailed location information)
```

Also, I added support for specifying an attrset directly in the repos default.nix, so this works too now:
```nix
{
  lib = import ./lib;
  modules = import ./modules;
}
```

I'll put the logic to evaluate a repository into a separate Nix file in future commits, such that the update script won't have to duplicate it.

Mic92 commented on default.nix:
```diff
@@ -1,7 +1,8 @@
-
-{ pkgs ? import <nixpkgs> {} }:
+{ nurpkgs ? import <nixpkgs> {} # For nixpkgs dependencies used by NUR itself
+, pkgs ? import <nixpkgs> {} # Dependencies to call NUR repos with
```
Should we make this argument mandatory and add a error message if not supplied by the user?

Mic92 replied:
The error message could then also include the canonical syntax to import NUR.

Infinisil replied:
Not sure, making the `pkgs` argument mandatory would mean that `import <nur> {}` wouldn't work, and you couldn't evaluate it anymore directly, e.g. `nix-instantiate '<nur>' -A repos.infinisil.foo` wouldn't work anymore.

Mic92 replied:
Should it then default to one `import` expression for example:

```
let fallback = import <nixpkgs> {} in {
  nurpkgs ?  fallback,
  pkgs ? fallback
}
```

I am not sure if this make any difference in terms of performance.

Infinisil replied:
According to @cleverca22 and @LnL7 in #nixos, what you proposed should be faster, so I'll change it to that.

Mic92 commented:
+1 for adding warnings.

For the sake of simplicity I would prefer if all repo root expression take `pkgs` by default even if they only define modules. The reason is that repo owners might otherwise by accident use the common pattern:

```nix
with import <nixpkgs> {};
{
}
```

because they don't understand the implication of it yet.
I don't think

```
{ pkgs }: {
}
```

is too much boiler code to ask for and makes all modules looks the same (also simpler documentation).
I would appreciate if you also add documentation. I will then notify all repository owner to update there repository.

Infinisil commented:
Yeah that sounds like a good idea. Should we also disallow defaulted arguments though? E.g. `{ pkgs, lib ? pkgs.lib }: { ... }` is quiet useful.

Infinisil commented:
I need some help with python (I don't know python well at all): The latest commit needs the path to the new evalRepo.nix file in python, but I'm not sure how to get it.

Infinisil commented:
The update script is now working again with the changes I made. (For some reason eeva's repo is being evaluated every time, but that was like this before my changes as well).

Infinisil commented on nur/update.py:
```diff
@@ -169,7 +169,11 @@ def eval_repo(spec: RepoSpec, repo_path: Path) -> None:
             f.write(f"""
                     with import <nixpkgs> {{}};
-callPackages {repo_path.joinpath(spec.nix_file)} {{}}
+import /home/infinisil/src/NUR/evalRepo.nix {{
```
This here, I don't think it can stay like this :P

Mic92 commented on evalRepo.nix:
```diff
@@ -0,0 +1,20 @@
+{ name
```
I think to establish best practices, this could go to `lib/`

Mic92 commented on nur/update.py:
```diff
@@ -184,6 +188,7 @@ def eval_repo(spec: RepoSpec, repo_path: Path) -> None:
             "-I", f"nixpkgs={nixpkgs_path()}",
             "-I", str(repo_path),
             "-I", str(eval_path),
+            "-I", str("/home/infinisil/src/NUR/evalRepo.nix")
```
If everything would be in `lib` then `"-I", str(ROOT.joinpath("lib"))` would cover it.

Mic92 commented:
If you have trouble get it working, I can take tomorrow a look at it again.

Infinisil commented:
Thanks! Got it figured out, there was a problem of ROOT being just  `.` previously, it needed a `.resolve()` to make it absolute and therefore work correctly for `-I`

I'll add some docs in future commits, and will squash commits into reasonable changes when ready for merge

Mic92 commented on README.md:
```diff
@@ -23,9 +23,9 @@ To make NUR accessible for your login user, add the following to `~/.config/nixp
-    nur = pkgs.callPackage (import (builtins.fetchGit {
-      url = "https://github.com/nix-community/NUR";
-    })) {};
+    nur = import (builtins.fetchGit "https://github.com/nix-community/NUR") {
```
I think it is also time to switch to `fetchTarball`, which is apparently faster then `fetchgit` - even for incremental updates - since the repository is rather small.

Infinisil replied:
Yeah, will be a bit longer though, but I think that's okay

Mic92 commented on README.md:
````diff
@@ -135,13 +135,13 @@ in stdenv.mkDerivation rec {
 
 ```console
-$ nix-shell -E 'with import <nixpkgs>{}; (callPackage ./default.nix {}).inxi'
+$ nix-shell -E '(import ./default.nix { pkgs = import <nixpkgs>{}; }).inxi'
````
Is the import required here? There is a default nixpkgs import in the file.

Mic92 replied:
Ah, sorry this is for the repository, not for NUR.

Mic92 commented on README.md:
````diff
@@ -135,13 +135,13 @@ in stdenv.mkDerivation rec {
 
 ```console
-$ nix-build -E 'with import <nixpkgs>{}; (callPackage ./default.nix {})'
+$ nix-build -E '(import ./default.nix { pkgs = import <nixpkgs>{}; }).inxi'
````
With the new repository format the following is also possible:

```console
nix-build --arg pkgs "import <nixpkgs> {}" -A inxi
```

Infinisil commented:
I added the `lib` argument to arguments a NUR repo receives because otherwise it would be hard to write other library functions without depending on pkgs, which would bring infinite recursion when those functions are used in certain ways in NixOS modules.

This PR is more-or-less ready now, probably needs some tweaks here and there.

zimbatm commented on README.md:
````diff
@@ -279,10 +306,12 @@ The result can be used like this:
 
 ```nix
-{ lib }:
+{ pkgs, lib }:
````
Is it necessary to pass `lib` if it's accessible through `pkgs.lib`?

Infinisil replied:
Using `lib` without a dependency on `pkgs` means that it's possible to use library functions (which are built only via `lib`) in your NixOS config without having to worry about infinite recursion, e.g. when using such a function in an overlay (using `pkgs.lib` would give inf rec because pkgs depends on overlays).

Mic92 replied:
In an overlay `lib` from `super` is supposed to be [used](https://blog.flyingcircus.io/2017/11/07/nixos-the-dos-and-donts-of-nixpkgs-overlays/) and for nixos modules, `lib` from the function argument is preferred. Would the current approach bypass overlays by reimporting `nixpkgs`? Are there any other use cases that I have not covered yet?

Infinisil commented on README.md:
```diff
@@ -211,49 +224,63 @@ To fetch git submodules in repositories set `submodules`:
-  modules = ./import modules;
+{ pkgs, config, lib, ... }:
+let
+  nur = import (builtins.fetchTarball "https://github.com/nix-community/NUR/archive/master.tar.gz") {};
```
Hmm, I'm not sure, because `pkgs` isn't actually needed when you just use NUR for modules/overlays/library functions.

Oh I think what would be neat is to change the default `pkgs` argument used by NUR from `import <nixpkgs> {}` to `throw "NUR didn't receive a pkgs argument"`. Then usage without a `pkgs` argument is just fine as long as you don't evaluate it (which is the case when you only need NixOS modules), but as soon as you try to use a package, you get the throw.

Infinisil commented on default.nix:
```diff
@@ -1,7 +1,13 @@
-{ pkgs ? import <nixpkgs> {} }:
 let
-  inherit (pkgs) fetchgit fetchzip callPackages lib;
+  inherit (nurpkgs) fetchgit fetchzip lib;
```
Probably not at this line of code, but yeah I can add that to the readme :)

Mic92 replied:
Is possible to prevent garbage collection by linking used tarballs in a profile.

Infinisil commented:
I rebased on master, and squashed all commits into 3, one for docs, one for nix and one for python.

Infinisil commented on default.nix:
```diff
@@ -1,7 +1,10 @@
+{ nurpkgs ? import <nixpkgs> {} # For nixpkgs dependencies used by NUR itself
+  # Dependencies to call NUR repos with
+, pkgs ? throw "NUR call didn't receive a pkgs argument, but the evaluation requires it."
```
This is new, I think this should be okay, there's no need to provide an implicit pkgs

Infinisil replied:
Also coincides with requiring a `pkgs` before this PR, as it was `pkgs.callPackage (fetchGit ...)`

Mic92 replied:
This change is great because it will help users when using NUR wrong.

Infinisil replied:
I can add a note that says it could be an error on either side.

Mic92 commented on README.md:
````diff
@@ -73,19 +88,39 @@ for its content.
+```nix
+{ pkgs, config, lib, ... }:
+let
+  nur = import (builtins.fetchTarball "https://github.com/nix-community/NUR/archive/master.tar.gz") {};
````
It is easy to shadow nur by using `environment.systemPackages = [ nur.repos.mic92.inxi ];` instead of `environment.systemPackages = [ pkgs.nur.repos.mic92.inxi ];`. 
In this case `pkgs` in NUR will be undefined and the error message you added is thrown.
Maybe the example should not called it `nur` to avoid this namespace collision?
The alternative could be also to make `nur` a nixos module (also it is probably not possible to dynamically inject imports and overlays?) - maybe a function that returns a nixos module could work.

Infinisil replied:
> The alternative could be also to make nur a nixos module

Yeah I thought of that! Would actually solve this problem rather elegantly: Users could add something like `imports = [ (import <nur/modules> {}) ]` (replace nur with the fetch command) exactly once to their nixos config.

To make this work, NUR should wrap every repositories modules under an enable option. So e.g. when I now want to use your foo module (which you declared in `modules.foo`), I'd do `nur.mic92.foo.enable = true`. This would work like this, and I might implement this in a future PR. For now this more hacky solution should be fine, especially since it can be forwards compatible.

Infinisil replied:
Oh and regarding the `nur` attribute name, I think changing it to `nur_for_modules` would work well.

Mic92 replied:
True, this can be also implemented in future. I also intentionally put repos as sub namespace so we can put more into the top-level hierarchy.

Mic92 replied:
Tracked here: https://github.com/nix-community/NUR/issues/43

Mic92 replied:
`s/nur_for_modules/nur-modules/g`

Infinisil replied:
Maybe `nur-no-pkgs`? Because the example there also uses overlays with it.

Mic92 replied:
Agreed.

Mic92 commented:
Sorry for being extra nitpicky and critical on this pull request, but since it requires changes for both users and maintainers I would like to keep the revision count on the API low.

Mic92 commented on README.md:
```diff
@@ -35,13 +35,28 @@ For NixOS add the following to your `/etc/nixos/configuration.nix`:
 
+### Pinning
+
+Using `builtins.fetchTarball` without a sha256 will only cache the download for 1 hour by default, so you need internet access almost every time you build something. You can pin the version if you don't want that:
```
**Update** I have not yet understand the full consequence of a fixed-input derivation: I have a non-stable URL in configuration.nix that I gave a checksum and nix is testing it from time to time and complains if the checksum changes. On the other hand I never had evaluation errors because the file is not available. Does this only works in my case because the referenced file is part of system profile closure or does this also apply to imported nix expression?

Infinisil replied:
For fixed output derivations (ones you give a hash), Nix will first check if a file with that hash is already in the store, and use that if there is. It will only try to fetch it when no such hash could be found. After the fetch it will only succeed in adding the download if the hash you gave matches the one it calculated. If it doesn't match, it fails. The fetchers from nixpkgs will also output the correct hash on a mismatch.

Mic92 replied:
How does nix decide, when to delete fixed-input derivations in this case?

Infinisil replied:
I think the only way it deletes them is garbage collection.

Infinisil commented:
> Sorry for being extra nitpicky and critical on this pull request

No problem, I'm also always very much like this on non-trivial nixpkgs PR reviews :)

Infinisil commented:
Added a super descriptive error message which includes the repository the error originates from.

Mic92 commented:
Feel free to merge this, if you think it is ready. Please also notify all repo users so they can updated accordingly.

Infinisil commented:
Made some minor adjustments and fixes:
- lib doesn't get passed to the repo, people can use pkgs.lib. We can pass more arguments later if the need arises (it probably won't)
- Updated python script to latest changes
- Having the "`pkgs` not passed" warning work with the deprecated syntax.

I tested everything now and I think it's ready. I'll merge this today/tomorrow.

zimbatm commented on README.md:
```diff
@@ -23,9 +23,9 @@ To make NUR accessible for your login user, add the following to `~/.config/nixp
-    nur = pkgs.callPackage (import (builtins.fetchGit {
-      url = "https://github.com/nix-community/NUR";
-    })) {};
+    nur = import (builtins.fetchTarball "https://github.com/nix-community/NUR/archive/master.tar.gz") {
```
Sorry I missed why using callPackage was an issue. Both are fine for me

Infinisil replied:
Because when you do `pkgs.callPackage`, this needs to evaluate `pkgs` (to get the attribute), which requires evaluating all overlays and the config (because they're passed to the nixpkgs import`, but those overlays can be defined in all modules. So if you want to `imports = [ nur.repos.infinisil.modules.foo ]`, this gives infinite recursion, because to evaluate the modules it needs to evaluate pkgs!

zimbatm replied:
Thanks, that makes a lot of sense

fgaz commented:
Sorry for saying this after the merge, but I think packages should have been in a separate attribute too, since now there's no way to only get packages from a NUR without doing [this](https://github.com/fgaz/nur-packages/blob/master/overlay.nix#L18)

Infinisil commented:
@fgaz Yeah I thought of that too. Doesn't really have anything to do with this PR though, it was the same before. Maybe open an issue regarding that

tilpner commented:
@fgaz [I have a pkgs attribute in my nur-packages](https://github.com/tilpner/nur-packages/blob/35bd43bf7bc6419a0035827a5c04be57b4d281bc/default.nix#L11), additional to having them as top-level attributes. I use that to overlay just my packages locally (lib would cause problems if overlayed naively).

Do you intend for something like this to become convention, or do you want all top-level packages to be gone?

fgaz commented:
@tilpner yeah that's definitely better than my filter...

I still think the top level should only be the four sets `pkgs`, `lib`, `modules`, and `overlays` though.

I opened an issue about it: #51
